cmd/proofgen/main.go

```diff
@@ -0,0 +1,511 @@
+		return nil
+	}
+
+	probes := corruptConsistencyProof(size1, size2, root1, root2, proof)
+	for _, p := range probes {
+		writeConsistencyProbe(directory, p)
+	}
+
+	return nil
+}
+
+func writeStaticConsistencyTestData(directory string) error {
+	root1 := []byte("don't care 1")
+	root2 := []byte("don't care 2")
+	proof1 := [][]byte{}
+	proof2 := [][]byte{sha256EmptyTreeHash}
+
+	tests := []consistencyProbe{
+		{0, 0, root1, root2, proof1, "sizes are equal (zero) but roots are not", true},
+		{1, 1, root1, root2, proof1, "sizes are equal (one) but roots are not", true},
+		// Sizes that are always consistent.
+		{0, 0, root1, root1, proof1, "sizes are equal and proof is not empty where both sizes are zero", false},
+		{0, 1, root1, root2, proof1, "size1 is zero and does not equal size2", true},
+		{1, 1, root2, root2, proof1, "sizes are equal and proof is not empty where both sizes are one", false},
+		// Time travel to the past.
+		{1, 0, root1, root2, proof1, "size1 is greater than size2", true},
+		{2, 1, root1, root2, proof1, "size1 is greater than size2 again", true},
+		// Empty proof.
+		{1, 2, root1, root2, proof1, "sizes do not watch and proof is empty", true},
+		// Roots don't match.
+		{0, 0, sha256EmptyTreeHash, root2, proof1, "roots do not match and sizes are zero", true},
+		{1, 1, sha256EmptyTreeHash, root2, proof1, "roots do not not match and sizes are one", true},
+		// Sizes match but the proof is not empty.
+		{0, 0, sha256EmptyTreeHash, sha256EmptyTreeHash, proof2, "sizes match but proof is not empty and sizes are zero", true},
+		{1, 1, sha256EmptyTreeHash, sha256EmptyTreeHash, proof2, "sizes match but proof is not empty and sizes are one", true},
+		// Fail to validate empty tree
+		{0, 1, sha256EmptyTreeHash, sha256EmptyTreeHash, proof2, "consistency check on empty tree (size1 is zero) is useless", true},
+	}
+
+	for _, p := range tests {
+		err := writeConsistencyProbe(directory, p)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func writeConsistencyProbe(directory string, probe consistencyProbe) error {
+	fileName := strings.Replace(probe.Desc, " ", "-", -1) + ".json"
+
+	probeJson, err := json.Marshal(probe)
+	if err != nil {
+		return fmt.Errorf("Error marshaling probe: %s", err)
+	}
+
+	fileLocation := filepath.Join(directory, fileName)
+	err = os.WriteFile(fileLocation, probeJson, 0644)
+	if err != nil {
+		return fmt.Errorf("Error writing probe: %s: %s", fileName, err)
+	}
+
+	return nil
+}
+
+// extend explicitly copies |proof| slice and appends |hashes| to it.
+func extend(proof [][]byte, hashes ...[]byte) [][]byte {
+	res := make([][]byte, len(proof), len(proof)+len(hashes))
+	copy(res, proof)
+	return append(res, hashes...)
+}
+
+// prepend adds |proof| to the tail of |hashes|.
+func prepend(proof [][]byte, hashes ...[]byte) [][]byte {
+	return append(hashes, proof...)
+}
+
+func dh(h string, expLen int) []byte {
+	r, err := hex.DecodeString(h)
+	if err != nil {
+		panic(err)
```

Comment:
`log.Fatalf` for consistency (also below)

Reply:
Done 🤝